compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPadding.cpp

```diff
@@ -0,0 +1,273 @@
+
+    SmallVector<OpFoldResult> newOffsets(newMixedSizes.size(),
+                                         rewriter.getIndexAttr(0));
+    SmallVector<OpFoldResult> newStrides(newMixedSizes.size(),
+                                         rewriter.getIndexAttr(1));
+    SmallVector<int64_t> newStaticDims;
+    SmallVector<Value> newDynamicDims;
+    dispatchIndexOpFoldResults(newMixedSizes, newDynamicDims, newStaticDims);
+
+    Location loc = loadOp.getLoc();
+    Value newLoad = rewriter.create<IREE::Flow::DispatchTensorLoadOp>(
+        loc, adaptor.getSource(), newDynamicDims, newOffsets, newMixedSizes,
+        newStrides);
+    auto extractType = RankedTensorType::get(boundTensorType.getShape(),
+                                             boundTensorType.getElementType());
+    SmallVector<OpFoldResult> extractSizes = getMixedValues(
+        boundTensorType.getShape(), loadOp.getSourceDims(), rewriter);
+    rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(
+        loadOp, extractType, newLoad, newOffsets, extractSizes, newStrides);
+    return success();
+  }
+};
+
+/// Pattern to convert `flow.dispatch.tensor.store` operation when
+/// materializing the encoding. We create a larger empty tensor for the
+/// destination and insert the value into it. This way we do not create partial
+/// stores prematurely, which would be difficult to undo later on.
+struct MaterializeFlowDispatchTensorStoreOp final
+    : OpMaterializeEncodingPattern<IREE::Flow::DispatchTensorStoreOp> {
+  using OpMaterializeEncodingPattern::OpMaterializeEncodingPattern;
+
+  LogicalResult
+  matchAndRewrite(IREE::Flow::DispatchTensorStoreOp storeOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Only handle operations where the store covers the entire
+    // `!flow.dispatch.tensor` type.
+    if (!storeOp.isStoreToWholeTarget()) {
+      return rewriter.notifyMatchFailure(storeOp, "unhandled partial stores");
+    }
+
+    auto targetType = storeOp.getTargetType();
+    auto boundTensorType = cast<RankedTensorType>(targetType.getBoundType());
+    auto &typeConverter =
+        *getTypeConverter<MaterializePadEncodingTypeConverter>();
+    auto paddedType =
+        typeConverter.convertType<RankedTensorType>(boundTensorType);
+    if (paddedType == boundTensorType) {
+      return rewriter.notifyMatchFailure(storeOp, "bound type already valid");
+    }
+    if (paddedType.getShape() == boundTensorType.getShape()) {
+      // Let the Nop pattern handle this.
+      return rewriter.notifyMatchFailure(storeOp, "no padding applied");
+    }
+
+    Location loc = storeOp.getLoc();
+    auto valueType = cast<RankedTensorType>(adaptor.getValue().getType());
+    SmallVector<Value> dynamicResultSizes;
+    for (size_t dim = 0, e = valueType.getNumDynamicDims(); dim != e; ++dim) {
+      dynamicResultSizes.push_back(
+          rewriter.create<tensor::DimOp>(loc, adaptor.getValue(), dim));
+    }
```

Comment:
the `Flow::DispatchTensorStoreOp`'s list of sizes will include the dynamic sizes, or you could get the dynamic sizes from the dispatch binding the store is storing to.

Reply:
Thanks, I didn't realize these are mixed sizes